### PR TITLE
MGMT-9539: Add presigned iPXE script URL endpoint

### PR DESCRIPTION
**Pull request by carbonin**
### Add an endpoint which generates a pre-signed URL to the iPXE script for an infra-env.

The new endpoint itself is rather simple. It uses the URL builder to create the url and signs it using the same method we use for ISO downloads.

The more complicated parts are the two initial refactoring commits:

#### Deduplicate presigned URL swagger objects
Previously there were two url types defined in the swagger spec `presigned` and `infra-env-image-url` both served the same purpose and had a `url` attribute.
This commit combines these implementations into a new type, `presigned-url` which will also be used for the upcoming endpoint to get a url to download the ipxe boot script.

#### Allow image auth in a query param
This commit adds a new security definition `imageURLAuth` to handle query parameter based auth for downloading infra-env files.
This is how a pre-signed URL for the ipxe-script will be implemented.
The authentication for this definition will be implemented using the same functions as imageAuth. The only differences here are how the token is provided (query param rather than header) and to which endpoints it applies (only the infra-env files download for url-based).

## List all the issues related to this PR

https://issues.redhat.com/browse/MGMT-9539

- [x] New Feature 
- [ ] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @filanov 
/cc @avishayt 
/cc @CrystalChun 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

### CODE REVIEW

File: swagger.yaml [hunk 1]
```diff
@@ -9129,8 +9173,10 @@ definitions:
         type: string
         format: uuid
 
-  infra-env-image-url:
```
Comment (filanov) [inline swagger.yaml:9132]: does it affect any of the clients?

File: internal/bminventory/inventory_v2_handlers.go [hunk 2]
```diff
@@ -643,3 +643,37 @@ func (b *bareMetalInventory) infraEnvIPXEScript(ctx context.Context, infraEnv *c
 
 	return fmt.Sprintf(ipxeScriptFormat, kernelURL, rootfsURL, initrdURL), nil
 }
+
+func (b *bareMetalInventory) GetInfraEnvIPXEScriptDownloadURL(ctx context.Context, params installer.GetInfraEnvIPXEScriptDownloadURLParams) middleware.Responder {
```
Comment (filanov) [inline internal/bminventory/inventory_v2_handlers.go:647]: Besides getting the URL is it going to be stored? 
I don't remember this part and i didn't find it in the enhancement

File: swagger.yaml [hunk 3]
```diff
@@ -4470,6 +4475,53 @@ paths:
           schema:
             $ref: '#/definitions/error'
 
+  /v2/infra-envs/{infra_env_id}/downloads/ipxe-script-url:
```
Comment (avishayt) [inline swagger.yaml:4478]: We already have:
/v2/clusters/{cluster_id}/downloads/files-presigned
So maybe create
/v2/infra-envs/{infra_env_id}/downloads/files-presigned
?

File: internal/bminventory/inventory_v2_handlers.go (hunk 2 above)
Comment (carbonin) [reply internal/bminventory/inventory_v2_handlers.go:647]: > it going to be stored?

No, only in the infraenv kube-api.

File: swagger.yaml (hunk 1 above)
Comment (carbonin) [reply swagger.yaml:9132]: I'm not sure, good question. I hope not as the API isn't actually changing, but maybe it would change the type name that was generated or something.

I don't think the agent is getting pre-signed URLs and given that the tests passed I guess it's not a problem for test-infra. Can we ask someone on the UI team?

File: swagger.yaml (hunk 3 above)
Comment (carbonin) [reply swagger.yaml:4478]: :shrug: Sure.

Do we care about making it work for the discovery ignition as well as the ipxe script?

File: swagger.yaml (hunk 1 above)
Comment (jkilzi) [reply swagger.yaml:6046]: **It will change the name of the generated TypeScript interface**

TL;DR: no big deal because the object remains the same (same members) it won't affect the UI code during runtime. In our dev-environments, we'll have to rename the interface in a few places.

PS: thanks for marking the `url` as required :)

**Review by filanov — APPROVED** [review]

**Comment by filanov** [comment]
was merged to 2.5 no need to backport.
